### PR TITLE
GEOS 3.9

### PR DESCRIPTION
**Pull request by scsides**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [ ] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [ ] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/base/objs/ImagePolygon/ImagePolygon.cpp [hunk 1]
```diff
@@ -240,7 +240,9 @@ namespace Isis {
         p_pts = NULL;
         p_pts = new geos::geom::CoordinateArraySequence();
 
+        std::cout << "C 1 -------------" << std::endl;
```
Comment (tgiroux) [inline isis/src/base/objs/ImagePolygon/ImagePolygon.cpp:243]: accidental cout

**Comment by Kelvinrr** [comment]
We can probably remove the "DO NOT MERGE" and make this a draft.

**Comment by jessemapel** [comment]
Status update from last time. Everything is looking good except for the isisminer overlap geometry tests. When one image completely overlaps another image it is showing up as non-overlapping now. I'm not sure why this is happening, but it's a serious issue that we need to figure out.

https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/src/base/apps/isisminer/GisOverlapStrategy.cpp#L203

This part of the code uses the GEOS STRtree

http://geos.refractions.net/ro/doxygen_docs/html/classgeos_1_1index_1_1strtree_1_1STRtree.html

I'm not seeing anything about functionality being different in the release notes, but it appears to have been worked on for the 3.9 release.

https://github.com/libgeos/geos/blob/main/NEWS#L39

File: isis/src/base/objs/PolygonTools/PolygonTools.cpp [hunk 2]
```diff
@@ -2148,6 +2163,7 @@ namespace Isis {
           }
           newLonLatPts->add(geos::geom::Coordinate(lon, lat), k);
 
+          delete pts3;
```
Comment (tgiroux) [inline isis/src/base/objs/PolygonTools/PolygonTools.cpp:2166]: this causes a segfault when the for loop checks condition after first iteration.

**Review by tgiroux — COMMENTED** [review]
see line 112-126 of ProcessPolygons.cpp.

This is apparently in dev but it looks like a mistake. The same try/catch block is duplicated back-to-back

**Comment by KrisBecker** [comment]
> Status update from last time. Everything is looking good except for the isisminer overlap geometry tests. When one image completely overlaps another image it is showing up as non-overlapping now. I'm not sure why this is happening, but it's a serious issue that we need to figure out.
> 
> https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/src/base/apps/isisminer/GisOverlapStrategy.cpp#L203

Have you checked the polygons for validity? It may simply be that one or more geometries are not valid.

**Comment by AustinSanders** [comment]
I wasn't able to solve either of the remaining challenges (the segfault or the polygon overlap issue) before the end of the fiscal. Some comments for anyone who may be picking up this issue:

I had trouble building directly off of the upstream/geos branch.  In order to get this built, I had to edit point_types.hpp with the changes listed [here](https://github.com/PointCloudLibrary/pcl/pull/3237).  Specifically, I edited the .hpp that was installed into my conda environment.

I wasn't able to make much headway on the segfault issue, but it does not appear to segfault reliably / deterministically.  This will likely require some pretty dedicated digging.

Regarding the overlapping geometries, I wasn't able to isolate any specific cause here.  I validated the geometries  [via python](https://gist.github.com/AustinSanders/7c60224428d1bfe8031913f88ec6feeb).

**Comment by jlaura** [comment]
@scsides Can we get this merged into a GEOS branch over here? If we don't and you delete the upstream, all of this work is lost.

**Pull request closed without merging**